sensor/common/sensor/central_communication_test.go

```diff
@@ -142,6 +136,105 @@ func (c *centralCommunicationSuite) Test_StopCentralCommunication() {
 	}
 }
 
+func (c *centralCommunicationSuite) Test_ClientReconciliation() {
+	dep1 := givenDeployment(fixtureconsts.Deployment1, "dep1", map[string]string{"app": "central"})
+	dep2 := givenDeployment(fixtureconsts.Deployment2, "dep2", map[string]string{"app": "sensor"})
+	updatedDep1 := givenDeployment(fixtureconsts.Deployment1, "dep1", map[string]string{"app": "central_updated"})
+	hasher := hash.NewHasher()
+	hash1, _ := hasher.HashEvent(dep1.GetEvent())
+	updatedHash1, _ := hasher.HashEvent(updatedDep1.GetEvent())
+	setSensorHash(dep1, hash1)
+	setSensorHash(updatedDep1, updatedHash1)
+	testCases := map[string]struct {
+		deduperState           map[string]uint64
+		componentMessages      []*central.MsgFromSensor
+		neverMatchesState      func(messages []*central.MsgFromSensor) bool
+		eventuallyMatchesState func(messages []*central.MsgFromSensor) bool
+	}{
+		"Deduper hash hit": {
+			deduperState: map[string]uint64{
+				deploymentKey(fixtureconsts.Deployment1): hash1,
+			},
+			componentMessages: []*central.MsgFromSensor{dep1},
+			neverMatchesState: anyDeploymentSent,
+		},
+		"All deployments sent": {
+			deduperState:           map[string]uint64{},
+			componentMessages:      []*central.MsgFromSensor{dep1, dep2},
+			eventuallyMatchesState: deploymentsSentCount(2),
+		},
+		"Updated deployment": {
+			deduperState: map[string]uint64{
+				deploymentKey(fixtureconsts.Deployment1): hash1,
+			},
+			componentMessages:      []*central.MsgFromSensor{updatedDep1},
+			eventuallyMatchesState: deploymentIDSent(fixtureconsts.Deployment1, updatedHash1),
+		},
+	}
+
+	for name, tc := range testCases {
+		c.Run(name, func() {
+			responsesC, closeFn := c.createCentralCommunication(true)
+			defer closeFn()
+			syncMessages := append(centralSyncMessages, debuggerMessage.DeduperState(tc.deduperState))
+			expectSyncMessages(syncMessages, c.mockService)
+
+			var sentMessages []*central.MsgFromSensor
+			mu := &sync.Mutex{}
+			c.mockService.client.EXPECT().Send(gomock.Any()).AnyTimes().DoAndReturn(func(msg *central.MsgFromSensor) error {
+				mu.Lock()
+				defer mu.Unlock()
+				sentMessages = append(sentMessages, msg)
+				return nil
+			})
+			c.mockService.client.EXPECT().CloseSend().AnyTimes()
+
+			reachable := concurrency.Flag{}
+			// Start the go routine with the mocked client
+			go c.comm.(*centralCommunicationImpl).sendEvents(c.mockService, &reachable, c.mockHandler, c.mockDetector, c.comm.(*centralCommunicationImpl).receiver.Stop, c.comm.(*centralCommunicationImpl).sender.Stop)
+			c.mockService.connected.Wait()
+
+			for _, msg := range tc.componentMessages {
+				responsesC <- message.New(msg)
+			}
+
+			if tc.eventuallyMatchesState != nil {
+				c.Assert().Eventually(func() bool {
+					mu.Lock()
```

Comment:
I tried both suggestions and I was still getting data races. I think it is because reading the slice to pass it to the function is already a read outside of the lock. Maybe I missed something though, let's move the conversation to #8084.